crypto/dh/dh_check.c

```diff
@@ -257,9 +257,27 @@ int ossl_dh_check_priv_key(const DH *dh, const BIGNUM *priv_key, int *ret)
     two_powN = BN_new();
     if (two_powN == NULL)
         return 0;
-    if (dh->params.q == NULL)
+
+    if (dh->params.q != NULL) {
+        upper = dh->params.q;
+#ifndef FIPS_MODULE
+    /*
+     * We do not have q so we just check the key is within the range
+     * we generate it.
+     */
+    } else if (dh->params.p != NULL) {
+        int32_t length = dh->length;
+
+        if (length == 0)
+            length = BN_num_bits(dh->params.p) - 1;
+        if (BN_num_bits(priv_key) <= length
+            && BN_num_bits(priv_key) > 1)
```

Comment:
I think it's only ok if BN_num_bits(priv_key) == length

Reply:
Nope, any key within reasonable range should be OK, there is no point in being overly strict.

Reply:
I think that if standards exists, we should follow them unless there is a good reason not to follow them. And I think we should be clear about which standards we are following.

Reply:
Which standard do you have on mind. Our DH implementation supports multiple of them and some of them (PKCS#3) is even more lenient about the private key value than this.

Reply:
PKCS#3 requires that if we have a length, that it should be that length. If we don't have a length, it should be 0 < x < p−1.

Reply:
It's not. The code I expect it something like:
```
if ((dh->length != 0 && BN_num_bits(priv_key) == dh->length)
|| (dh->length == 0 && BN_num_bits(priv_key) <= (BN_num_bits(dh->params.p) - 1)  && BN_num_bits(priv_key) > 1) {
ok = 1
}
```

Reply:
Fixup pushed